cli_utils/unpackdat.c

```diff
@@ -90,47 +140,105 @@ void close_dat(struct datfile *rf) {
 	return;
 }
 
-#define DCHUNK_SIZE 0x8D	//chunk size is very happy
-#define DCHUNK_DLEN 0x80	//bytes of payload
-// "FMT SRC TGT LEN 34 82 AH AM AL CL <CL bytes of data> <CRC1> <CRC2> <CKS>"
-void unpack_dat(FILE *outf, uint8_t *dest, const uint8_t *src, u32 siz) {
-	// 1- search for "88 34 82"
-	const uint8_t pm[]={0x88, 0x34, 0x82};
+void print_hex(const void *src, int len) {
+	const uint8_t *ptr = src;
+	for (int i = 0; i < len; i++) {
+		printf(" %02x", ptr[i]);
+	}
+	printf("\n");
+}
+
+void unpack_dat(FILE *outf, struct datfile *rf, bool trim) {
+	const uint8_t *src = rf->buf;
+	const u32 src_size = rf->buf_size;
+	const uint8_t pm[] = {0x34, 0x82};
 	const uint8_t *psrc;	//misc pointer inside src buf
-	uint8_t *orig_dest = dest;
-	uint32_t curaddr = 0;
+	uint8_t *dest = NULL;
+	
+	const chunk_header_t *ch_header = NULL;
 
-	unsigned long plsiz = 0;	//total payload bytes
+	uint8_t chklen = 0;
+	uint32_t offset = 0;
+	uint32_t firstchaddr = 0;
+	unsigned long outsize = 0;
 	unsigned pl_numc = 0;	//# of chunks
 
-	psrc = u8memstr(src, siz, pm, 3);
-	if (!psrc) return;
-
-	//2- point to "AH AM AL..." and start
-	psrc += 3;
-	while ( psrc < &src[siz - 1] ) {
-		// parse AH:AM:AL
-		uint32_t temp_addr = reconst_32(psrc);
-		temp_addr >>= 8;
-		if (temp_addr != (curaddr + DCHUNK_DLEN)) {
-			printf("addr skip @ %lX\n", (unsigned long) temp_addr);
+	const dat_header_t *dath = (const dat_header_t *)src;
+
+	printf("SIGN:");
+	print_hex(dath->sign, 3);
+	printf("PART1: %.*s\n", 5, dath->part_number);
+	printf("DATE:");
+	print_hex(&dath->date, sizeof(dath->date));
+	printf("PART2: %.*s", 5, dath->part_number2);
+	print_hex(dath->part_number2, 5);
+	printf("VER: %.*s", 6, dath->ver_str);
+	print_hex(dath->ver_str, 6);
+
+	// TODO get rid of search. The first chunk is located at a constant offset 0x25
+	psrc = u8memstr(src, src_size, pm, 2);
+	if (!psrc) {
+		printf("Pattern 34 82 not found\n");
+		return;
+	}
+	
+	ch_header = (const chunk_header_t *)(psrc - sizeof(chunk_header_t));
```

Comment:
here too, casting directly to struct.
And because you're backing up by a fixed size this may break with my weird 9HA7D.dat as I mentioned